Init function vs times from file

geoffrey4444 opened the pull request:
## Proposed changes

This PR adds code to read in FunctionOfTime data from an H5 Dat file. This is a step toward evolving BBHs using FunctionOfTime data from a corresponding spec simulation.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

geoffrey4444 commented:
Fixed my silly mistakes causing test failures (wrong case for domain namespace, forgot to remove a debugging include iostream). @kidder, unless there’s something else you’d like me to do first, I think this is ready for a look! I removed the in progress label, but I’m happy to reapply it if it was for something other than those things I just fixed first

geoffrey4444 committed:
Add action to read functions of time from file

kidder commented:
@geoffrey4444 I marked as `In progress` as it doesn't compile; please don't remove the label until you get the green check mark...

geoffrey4444 commented:
I just fixed a test I hadn’t realized was failing, and I refactored the commits to clarify changes to enable storing types in the data box that don’t support <<

geoffrey4444 commented:
Ok @kidder...I think I’ve got it fixed, will wait to be sure before removing that tag. Sorry for removing it prematurely! I didn’t realize there was another test that I’d broken, but now it should be fixed

hrueter commented on src/Domain/FunctionsOfTime/Tags.hpp:
```diff
@@ -0,0 +1,25 @@
+/// \ingroup ControlSystemGroup
+/// \ingroup DataBoxTagsGroup
+/// An unordered map of `FunctionOfTime` objects
+struct FunctionsOfTime : db::SimpleTag {
```
- Name this tag `PiecewisePolynomialFromFile` and template it on an `OptionGroup`. Then add a `create_from_options` function and this type alias

```cpp
using option_tags = tmpl::list<OptionTags::FunctionOfTimeFile<OptionGroup>,OptionTags::FunctionOfTimeDataSet<OptionGroup>>;
```

- Change the `type` to `PiecewisePolynomial`. 

- Move the relevant `OptionTags` from  src/Evolution/Initialization/Tags.hpp  into this file.

- Turn the function that creates the `PiecewisePolynomial` from the H5 data into a free function and call it in `create_from_options`. Then you don't need the action anymore.

geoffrey4444 replied:
Thanks for the very quick review!! I have two questions:

1. I made this an action so that evolution executables could have access to the read data. How can I use this in an executable if I make it a free function instead of an action? @nilsdeppe suggested I make this an action, so I'd like him to weigh in on this suggestion

2. @nilsdeppe told me, if I understood him correctly, that his moving mesh code stores the functions of time in an unordered map exactly as I have done here... @nilsdeppe, could you please weigh in on the suggestion to change this tag (and the code) to store individual functions of time in the databox instead of storing them all in an unordered map?

nilsvu replied:
1. With the `create_from_options` approach the `FunctionOfTime` also ends up in the DataBox and you can retrieve it from there in evolution executables. The way it gets into the DataBox is different: The option-creation mechanism constructs it instead of the action. You may still use an action to add it to the executable: Make an empty action that has both `initialization_tags` and `initialization_tags_to_keep` set to `tmpl::list<PiecewisePolynomialFromFile>`. If you prefer to have the action load the data instead of the option-creation mechanism (perhaps to have more control over _when_ the data is loaded, not sure if that's relevant) then I still suggest you turn the functionality into a free function and call it from the action.
2. Depends on how the functions of time are going to be used: Do we know at compile time which functions are needed? I.e. can we construct a struct like this

    ```cpp
    struct CorotationWithBinary {
      constexpr OptionString help = "A function of time for co-rotating the domain with an orbiting binary";
    };
    ```

    and use it to retrieve the function from the DataBox as `db::get<FunctionOfTime<CorotationWithBinary>>(box)`? Then my suggestion applies. If, on the other hand, we only know which functions are needed at runtime, then perhaps the string-indexed map is the way to go.

nilsdeppe replied:
Just leave this as is. It's the right design to work with the moving mesh code.

hrueter commented on src/Domain/FunctionsOfTime/Tags.hpp:
```diff
@@ -0,0 +1,25 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines tags related to FunctionOfTime quantities
```
You can remove the \file info.

hrueter commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+
+    // Currently, only support order 3 piecewise polynomials
+    // This could be generalized later, if needed
+    constexpr size_t max_deriv{3};
```
Is it straight forward enough to template this function on `MaxDeriv`?

nilsdeppe replied:
I guess technically, but there's no benefit to that since SpEC always uses 3

kidder replied:
@markscheel is this true?

geoffrey4444 replied:
@kidder spec sometimes uses second order, but I only need to read in 3rd order. I'll clarify in the code's name and documentation that this code reads 3rd order piecewise polynomials now. It can be generalized later in a future PR if needed

hrueter commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+      for (size_t i = 0; i < max_deriv + 1; ++i) {
+        gsl::at(initial_coefficients, i) = DataVector(number_of_components);
+        for (size_t a = 0; a < number_of_components; ++a) {
+          // Columns in the file have the following form::
```
remove the last `:`

hrueter commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+          // 2 = number of components
+          // 3 = maximum derivative order
+          // 4 = version
+          // 5, ... 5 + maximum derivative order: first component and derivs
```
This file structure should be part of the Doxygen documentation.
Line 106 was a little unclear for me. Please add a description and use a notation that is more consistent with the above columns. (`=`)

hrueter commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+      std::array<DataVector, max_deriv + 1> initial_coefficients;
+      for (size_t i = 0; i < max_deriv + 1; ++i) {
+        gsl::at(initial_coefficients, i) = DataVector(number_of_components);
+        for (size_t a = 0; a < number_of_components; ++a) {
```
Please rename `i` to `deriv_order` and `a` to `component`.

hrueter commented on src/ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp:
```diff
@@ -57,10 +57,7 @@ void merge_simple_tag_value(const gsl::not_null<db::DataBox<DbTagsList>*> box,
-             "value in the DataBox is: "
-          << db::get<SimpleTag>(*box) << " while the value being added is "
-          << simple_tag);
+          << " is not the same as what is already in the DataBox.");
```
Please revert this change. You don't need this anymore once you have factored the action into a free function.

nilsdeppe replied:
The way I think we should handle this is:
```cpp
namespace db {
template <typename T>
std::string get_tag_value(const T& t) noexcept {
  return get_output(t);
}

template <typename T>
std::string get_tag_value(const std::unique_ptr<T>& t) noexcept {
  if (t == nullptr) {
    return "nullptr";
  }
  return get_output(*t);
}
}  // namespace db
```
in a file `DataStructures/DataBox/GetTagValue.hpp` and then invoke that function here as `db::get_tag_value(db::get<SimpleTag>(*box))`. I can put this together into a PR quick if you like, @geoffrey4444.

geoffrey4444 replied:
OK, for now I'm going to leave this as-is, until we figure out what the real culprit is

hrueter reviewed (changes requested):
group review by @nilsleiffischer (main) and @hrueter .
We have not reviewed the tests yet.

In the comment below we suggest to store each function of time in a separate tag instead of in a string indexed map. Since the tags are distinguished by their `OptionGroup` template parameter the user can retrieved the one they need from the databox. They supply the `OptionGroup` that labels the function of time. Hopefully this works for the use case you have in mind for this functionality.

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+///
+/// DataBox changes:
+/// - Adds:
+///   - A `FunctionsOfTime` containing a `FunctionOfTime` for each dat
```
I think this should be `:domain::Tags::FunctionsOfTime` since we care about exactly what tags are being added. You can add the description after the tag name

geoffrey4444 replied:
Did you mean `::domain::Tags::FunctionsOfTime`?

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+  template <typename DataBox, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(DataBox& box,
```
DataBox -> `db::DataBox<DbTagsList>` to force type matching

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+struct FunctionOfTimeFromFile {
+  template <typename DataBox, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
```
Should have a `Requires` for the tags that are expected to be in the DataBox

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+    // element 0 is the dataset name in the h5 file, and element 1 is the
+    // name used in the std::unordered_map that will store the read-in functions
+    // of time
+    const auto& dataset_names{
```
If you make the type explicit `std::array<std::string, 2>` then the comment can be shortened and we have the compiler checking more things

geoffrey4444 replied:
Sure! Do you have a good feeling how I know when is a good idea to use `auto`? The explanation in the dev guide just says basically unless you have a `DataVector`, so I'd been defaulting to using auto when possible except for `DataVectors`

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& file_name{
```
`auto` -> `std::string` (for short type names I prefer to use the type name explicitly since it adds another level of check)

kidder replied:
`const std::string& file_name =` (and others below)

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+        functions_of_time;
+
+    h5::H5File<h5::AccessType::ReadOnly> file(file_name);
+    for (auto dataset_name : dataset_names) {
```
auto -> `const auto&`

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+
+    h5::H5File<h5::AccessType::ReadOnly> file(file_name);
+    for (auto dataset_name : dataset_names) {
+      const auto& dataset = dataset_name[0];
```
auto -> std::string (and on next line)

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+      const auto& dataset = dataset_name[0];
+      const auto& name = dataset_name[1];
+      const auto& dat_file = file.get<h5::Dat>("/" + dataset);
+      const auto& dat_data = dat_file.get_data();
```
`const auto&` -> `const Matrix`

Could you add a comment explaining what the rows and columns represent?

geoffrey4444 replied:
Added to help text

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+    for (auto dataset_name : dataset_names) {
+      const auto& dataset = dataset_name[0];
+      const auto& name = dataset_name[1];
+      const auto& dat_file = file.get<h5::Dat>("/" + dataset);
```
auto -> `h5::Dat`

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+      // Check that the data in the file uses deriv order 3
+      // Column 3 of the file contains the derivative order
+      const size_t dat_max_deriv = dat_data(0, 3);
+      ASSERT(dat_max_deriv == max_deriv,
```
Since this is a runtime check that could go bad in release mode (it depends on external files), let's make this an `if () { ERROR(); }`

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+          highest_derivative[a] =
+              dat_data(row, 4 + (max_deriv + 1) * a + max_deriv);
+        }
+        functions_of_time[name].update(dat_data(row, 1), highest_derivative);
```
I'm not sure what happens when we repeated insert the same time of last update. I was expecting this to only `update` when the `time_of_last_update` is changed in the file.

geoffrey4444 replied:
OK, I've changed this to skip rows where time of last update is not updated

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    // Create an unordered map of the same type as the FunctionsOfTime tag,
+    // and then add this to the databox
```
`databox` -> `DataBox`

geoffrey4444 replied:
Comment no longer present because now it's a mutating action, not an adding action

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+    }
+    return std::make_tuple(::Initialization::merge_into_databox<
+                           FunctionOfTimeFromFile,
+                           db::AddSimpleTags<::domain::Tags::FunctionsOfTime>>(
```
This should assume `FunctionsOfTime` is already in the DataBox and make sure the keys of the existing one match the keys of those being added. I'm happy to make that adjustment once the moving mesh code is done too, up to you.

geoffrey4444 replied:
I'll take a stab at this

nilsdeppe commented on src/Domain/FunctionsOfTime/Tags.hpp:
```diff
@@ -0,0 +1,25 @@
+#include <string>
+#include <unordered_map>
+
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
```
I think you need `DataStructures/DataBox/DataBoxTag.hpp`

kidder replied:
ping

nilsdeppe commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,145 @@
+/// - Removes: nothing
+/// - Modifies: nothing
+struct FunctionOfTimeFromFile {
+  template <typename DataBox, typename... InboxTags, typename Metavariables,
```
I think you want to list the initialization tags in a `using initialization_tags = tmpl::list<...>;` type alias

geoffrey4444 replied:
I'm confused by this comment...is it still relevant? If so, what are the initialization tags?

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
 #include "Utilities/TaggedTuple.hpp"
 
 namespace Initialization {
+namespace OptionTags {
```
Since these are really specific to the one action my vote would be to put them in the same file as the action or a `TagsReadingFunctionsOfTime.hpp`

kidder replied:
I agree, don't put these tags in this file

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
 
 namespace Initialization {
+namespace OptionTags {
+struct FuncOfTimeFromFile {
```
`FuncOf...` -> `FunctionOf...` If the option parser gives you grief about name length you can specify a shorter name using the `name()` function

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
+      "Options for reading a FunctionOfTime from an H5 file"};
+};
+
+struct FuncOfTimeFile {
```
`Func` -> `Function`

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
+
+struct FuncOfTimeFile {
+  using type = std::string;
+  static constexpr OptionString help{"Path to an H5 Dat file"};
```
I don't understand what this means. Do you mean the path to the HDF5 file or the path inside the HDF5 file to the H5 Dat subfile?

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
+  using group = FuncOfTimeFromFile;
+};
+
+struct FuncOfTimeSetNames {
```
`Func` -> `Function`

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
+struct FuncOfTimeSetNames {
+  using type = std::vector<std::array<std::string, 2>>;
+  static constexpr OptionString help{
+      "String pairs mapping datasets to arbitrary names"};
```
I don't know what info I am supposed to gain from this help string

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
 /// \brief %Tags used during initialization of parallel components.
 namespace Tags {
+/// \ingroup ControlSystemGroup
+/// \brief The path to an H5 Dat file containing function-of-time data to read.
```
again, I don't know what you mean here

nilsdeppe commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
+/// \ingroup ControlSystemGroup
+/// \brief List of pairs of strings mapping datasets to unordered-map keys
+///
+/// The first string in each pair names a dataset in an H5 file, and the
```
I don't know what a `dataset in an H5 file` is. That's not a data format we use.

geoffrey4444 replied:
I attempted to clarify...when you next take a look, please let me know if it makes sense this time

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,138 @@
+};
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Evolution.FunctionOfTimeFromFile",
```
Can you combine the two test cases by making the current test cases functions in an anonymous namespace and then calling them from the one `SPECTRE_TEST_CASE`? This will help reduce total test runtime

geoffrey4444 replied:
Sure! Good to know that this shortens the test runtime

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,138 @@
+struct component {
+  using metavariables = Metavariables;
+  using chare_type = ActionTesting::MockArrayChare;
+  using array_index = ElementIndexType;
```
you can just use `size_t` as well since I don't think the action assumes `ElementIndex`

geoffrey4444 replied:
When I tried this, I got these compiler errors. Changing back to ElementIndexType worked, so I'll leave it as that unless there's a fix you can suggest

~~~~~
/home/geoffrey/Codes/spectre/spectre/tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:91:3: error: no matching function for call to 'emplace_component_and_initialize'
  ActionTesting::emplace_component_and_initialize<component<Metavariables>>(
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/geoffrey/Codes/spectre/spectre/tests/Unit/ActionTesting.hpp:1623:6: note: candidate function not viable: no known conversion from 'const ElementId<dim>' to 'const typename component<Metavariables>::array_index' (aka 'const unsigned long') for 2nd argument
void emplace_component_and_initialize(
     ^
/home/geoffrey/Codes/spectre/spectre/tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:100:10: error: no matching member function for call to 'next_action'
  runner.next_action<component<Metavariables>>(self_id);
  ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/geoffrey/Codes/spectre/spectre/tests/Unit/ActionTesting.hpp:1538:8: note: candidate function not viable: no known conversion from 'const ElementId<dim>' to 'const typename component<Metavariables>::array_index' (aka 'const unsigned long') for 1st argument
  void next_action(
       ^
/home/geoffrey/Codes/spectre/spectre/tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:103:7: error: no matching function for call to 'get_databox_tag'
      ActionTesting::get_databox_tag<component<Metavariables>,
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/geoffrey/Codes/spectre/spectre/tests/Unit/ActionTesting.hpp:1668:13: note: candidate function not viable: no known conversion from 'const ElementId<dim>' to 'const typename component<Metavariables>::array_index' (aka 'const unsigned long') for 2nd argument
const auto& get_databox_tag(
~~~~~

kidder replied:
try again and change the type of `self_id` below...change `self_id` to `size_t`

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,138 @@
+  MockRuntimeSystem runner{{}};
+  ActionTesting::emplace_component_and_initialize<component<Metavariables>>(
+      &runner, self_id,
+      {db::item_type<Initialization::Tags::FuncOfTimeFile>{
```
Please write out the type

geoffrey4444 replied:
Sure! I got misled by the test I based this off of, to learn how the action framework works (can't remember which one) that did this

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,138 @@
+      {db::item_type<Initialization::Tags::FuncOfTimeFile>{
+           unit_test_path() +
+           "/Evolution/Initialization/FunctionOfTimeFromFile_TestData.h5"},
+       db::item_type<Initialization::Tags::FuncOfTimeSetNames>{
```
Please write out the type

nilsdeppe commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,138 @@
+                                     domain::Tags::FunctionsOfTime>(runner,
+                                                                    self_id);
+
+  // Check that the function vs. time and its derivatives have the expected
```
`function vs. time` -> `FunctionOfTime`

nilsdeppe reviewed (changes requested):
Overall looks good. Pretty much just minor changes :+1:

Please don't do @nilsleiffischer's refactoring change. That won't work with the moving mesh code.

geoffrey4444 commented:
> The way I think we should handle this is:
```cpp
namespace db {
template <typename T>
std::string get_tag_value(const T& t) noexcept {
  return get_output(t);
}

template <typename T>
std::string get_tag_value(const std::unique_ptr<T>& t) noexcept {
  if (t == nullptr) {
    return "nullptr";
  }
  return get_output(*t);
}
}  // namespace db
```
in a file `DataStructures/DataBox/GetTagValue.hpp` and then invoke that function here as `db::get_tag_value(db::get<SimpleTag>(*box))`. I can put this together into a PR quick if you like, @geoffrey4444.
 

Yes, would you mind please submitting this as a quick PR, @nilsdeppe? I'll review

wthrowe commented:
(Discussed with @nilsdeppe)  The proposed `get_tag_value` function above won't help, because `db::get` automatically dereferences `unique_ptr`s and the second overload will never be called.  There must be something else being added to the DataBox that is not streamable.  Our conclusion was that whatever it is should be made streamable, even if the stream operator cannot provide a full description of the class instance.  (At worst it could just print the class name.)

geoffrey4444 commented:
I think this is ready for another look...I got rid of the commit that let you add unstreamable functions to the DataBox and instead made a new commit that makes FunctionOfTime streamable

kidder commented on src/Domain/FunctionsOfTime/FunctionOfTime.hpp:
```diff
@@ -37,6 +37,21 @@ class FunctionOfTime : public PUP::able {
 
   WRAPPED_PUPable_abstract(FunctionOfTime);  // NOLINT
+
+  // Functions in the DataBox must be streamable, because operator<<() is
```
I do not understand the comment.  To my knowledge there is nothing that requires objects in the DataBox to be streamable.   I don't see stream operators for any of the other classes that are stored as `unique_ptr`s in the DataBox

geoffrey4444 replied:
I should amend to say "An object added as a `SimpleTag` in the DataBox must either be streamable itself or have support for streaming provided by `StdHelpers.hpp`. Otherwise, an error message in `merge_simple_tag_value` will fail to compile"

geoffrey4444 replied:
Note to self: when mutating instead of adding to databox, maybe don't need this. But if you do need it, make the stream operator virtual and print out entire internal state

geoffrey4444 replied:
Now that my action is a mutating action, this is no longer an issue, so I no longer am going to try to make FunctionOfTime streamable

kidder commented on src/Domain/FunctionsOfTime/Tags.hpp:
```diff
@@ -0,0 +1,22 @@
+#include <string>
+#include <unordered_map>
+
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
```
I think you just need a forward declaration of `::domain::FunctionsOfTime::FunctionOfTime` instead of this include.

kidder commented on src/Domain/FunctionsOfTime/Tags.hpp:
```diff
@@ -0,0 +1,22 @@
+      std::string, std::unique_ptr<::domain::FunctionsOfTime::FunctionOfTime>>;
+};
+}  // namespace Tags
+}  // namespace domain
```
why a new file?  I would just use `Domain/Tags.hpp`

geoffrey4444 replied:
I thought since it's in a different namespace, it should get a new file. But I'm fine moving it

kidder replied:
@nilsdeppe and I discussed this, and adding this file is fine

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+// Distributed under the MIT License.
```
I think this file should go in `IO/Importers` and be named `FunctionsOfTimeReader`

geoffrey4444 replied:
SpecFunctionOfTimeReader

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+namespace Initialization {
+namespace Actions {
+/// \ingroup InitializationGroup
+/// \brief Initialize a `FunctionsOfTime` from dat files in an
```
just say "from an H5 file."

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Initialization {
```
I would put this in the `importers` namespace, not `Initialization`

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+/// Uses:
+///  - DataBox:
+///    - `Initialization::Tags::FuncOfTimeFile`
+///    - `Initialization::Tags::FuncOfTimeSetNames`
```
What are set names?

geoffrey4444 replied:
See the help text below...they are the keys used to index the functions of time in the std::unordered_map they will be stored in

geoffrey4444 replied:
FuncOfTimeNames

geoffrey4444 replied:
FuncOfTimeNameMap, clarify spec names -> spectre names

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+///
+/// DataBox changes:
+/// - Adds:
+///   - `::domain::Tags::FunctionsOfTime`. This tag contains a `FunctionOfTime`
```
You don't need to comment after the tag name, it will be linked.  And shouldn't this be a mutating action not an initialization action?

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+    // element 0 is the dataset name in the h5 file,
+    // and element 1 is the name used in the std::unordered_map that
+    // will store the corresponding read-in function of time
+    const std::vector<std::array<std::string, 2>>& dataset_names{
```
why  are the names different?  why aren't they fetched from the existing FunctionsOfTime?

geoffrey4444 replied:
Because spec might use very un-spectre-like naming conventions. Also, re: your question to @markscheel, I should amend the comments. I now realize that the derive order is not always set to 3 in spec, but sometimes is set to 2 instead. For now, though, I only want to mimic a simple spec control system using max_deriv=3. If we end up wanting to mimic a more complex control system later, then this can be generalized to take max_deriv as an input option, but I wanted to keep the first version of this simple for now

kidder replied:
make this a std::map<std::string, std::string>

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+    const std::vector<std::array<std::string, 2>>& dataset_names{
+        db::get<Initialization::Tags::FuncOfTimeSetNames>(box)};
+
+    // Currently, only support order 3 piecewise polynomials
```
If this only supports one specific function of time, then the name of this action should be changed to reflect that

geoffrey4444 replied:
I plan to rename the action to `SpecFunctionOfTimeReader` and change the comment to read `Currently, only support order 3 piecewise polynomials. This could be generalized later, but the SpEC functions of time that we will read in with this action will always be 3rd-order piecewise polynomials.` Is that sufficient? Or would you prefer a name like `SpecThirdOrderPiecewisePolynomialReader`?

kidder replied:
We have been preferring action names to be verbs, so `ReadSpecThirdOrderPiecewisePolynomial`

kidder commented on src/Evolution/Initialization/FunctionOfTimeFromFile.hpp:
```diff
@@ -0,0 +1,164 @@
+
+      // Currently, assume the same number of components are used
+      // at each time. This could be generalized if needed
+      const size_t number_of_components = dat_data(0, 2);
```
generalize this now; pretty sure the shape map varies the number of components...
but wait doesn't that break the assumption the dataset is a Matrix???

geoffrey4444 replied:
Yes, I do not intend to support reading shape control at all (let alone varying the number of components) any time soon, because that's not a step on the way to the first BBH evolutions. I hope that by the time we're ready to do that, the spectre control system is ready to handle it.

Even if you insist that I generalize this "now," could that please be a separate PR, to avoid delaying my goal of getting to a BBH orbit ASAP (hopefully in time for APS)?

kidder commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
 /// \ingroup InitializationGroup
 /// \brief %Tags used during initialization of parallel components.
 namespace Tags {
+/// \ingroup ControlSystemGroup
```
wrong group

geoffrey4444 replied:
No group

kidder commented on src/Evolution/Initialization/Tags.hpp:
```diff
@@ -3,22 +3,80 @@
+  }
+};
+
+/// \ingroup ControlSystemGroup
```
wrong group

geoffrey4444 replied:
What is the right group, and where should the tags go?

geoffrey4444 replied:
No group

kidder commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,140 @@
+// Distributed under the MIT License.
```
why are we putting an H5 file in the repo; just have the test write it

geoffrey4444 replied:
Note to self: look at IO tests to see how to write a file, read it, and remove it
grep for remove_file()

kidder commented on tests/Unit/Evolution/Initialization/Test_FunctionOfTimeFromFile.cpp:
```diff
@@ -0,0 +1,140 @@
+      expected_functions;
+  expected_functions[expected_names[0]] = expected_rotation;
+  expected_functions[expected_names[1]] = expected_expansion;
+
```
please check that the size of `functions_of_time` is the expected size, as your test will pass if the fetched map is empty...

kidder replied:
`REQUIRE(functions_of_time.size() == expected_names.size());`

geoffrey4444 commented:
Note to self: add error handling that checks for non-monotonic times (decreasing or duplicated) and screams, and add test of that error handling code

geoffrey4444 commented:
I just pushed fixup commits and rebased on develop...I think this is ready for another look when you get a chance!

kidder commented:
@geoffrey4444  after you fix the compiler errors, please squash and we will take another look

geoffrey4444 commented:
Thanks @kidder! I'm baffled by the error, it was a segfault only on clang-9. But when I compiled with clang-9 in singularity, I couldn't reproduce. So I'm rerunning the tests to see what happens. If they pass, I'll squash

geoffrey4444 commented:
@kidder @nilsdeppe I'm unable to reproduce the test failure, but it seems related to my test of the run-time error catching non-monotonic time steps. Should I just remove that test so GitHub is happy?

kidder commented:
@geoffrey4444 in a commit before yours please add
`fail-fast: false` into `.github/workflows/Tests.yaml`  after line 225 (which says `strategy`) with the same indentation as `matrix` on the following line.  This will let us see which compilers it fails with

geoffrey4444 commented:
All tests except one fail, but when I try these compilers in my singularity environment on ocean, they all pass. The logs show that the failing test prints the ERROR() message as expected, but then segfaults for some reason. Any advice, @nilsdeppe and @kidder?

kidder commented:
start commenting out lines of the test until you find a minimal example that fails...

geoffrey4444 commented:
@kidder @nilsdeppe thanks to help from @moxcodes I have revised my ERROR() test so that it doesn't segfault. I think this is ready for another look when you get a chance

moxcodes commented:
I think you can drop the first of our segfault fix attempts and still use the action-testing framework in the error test in case that is preferable. I'm glad it seems to work!

geoffrey4444 commented:
> I think you can drop the first of our segfault fix attempts and still use the action-testing framework in the error test in case that is preferable. I'm glad it seems to work!

If possible I’d rather keep the error test as is, so we have an example of bypassing the action testing framework, for other times when that might be needed. I still use the action testing framework in the main test

kidder commented:
@geoffrey4444 actually I would like you to try to keep the action testing part, as that is one possibility I was worried might be the problem...

kidder reviewed (commented):
you can squash

geoffrey4444 commented:
@kidder @nilsdeppe this is ready for a look when you get a chance

nilsdeppe commented on src/IO/Importers/ReadSpecThirdOrderPiecewisePolynomial.hpp:
```diff
@@ -0,0 +1,195 @@
+          spec_functions_of_time[spectre_name].update(time_last_updated,
+                                                      highest_derivative);
+        } else {
+          file.~H5File();
```
Calling destructors explicitly is only okay if an object was allocated with placement new... Can you see if all you need to do is close the subfile? Otherwise you should open the file into a `unique_ptr` and then reset it to close the file

nilsdeppe commented on src/IO/Importers/ReadSpecThirdOrderPiecewisePolynomial.hpp:
```diff
@@ -0,0 +1,195 @@
+                domain::FunctionsOfTime::PiecewisePolynomial<max_deriv>*>(
+                (*functions_of_time)[spectre_name].get());
+            if (piecewise_polynomial == nullptr) {
+              ERROR("piecewise_polynomial should not be nullptr\n");
```
This error message wan't be helpful to users... How about `The function of time with name spectre_name is not a PiecewisePolynomial<max_deriv> and so cannot be set using ReadSpecThirdOrderPiecewisePolynomial`

nilsdeppe commented on src/IO/Importers/Tags.hpp:
```diff
@@ -142,6 +171,41 @@ struct RegisteredElements : db::SimpleTag {
+struct FunctionOfTimeFile : db::SimpleTag {
+  static std::string name() noexcept { return "FunctionOfTimeFile"; }
+  using type = std::string;
+  using option_tags = tmpl::list<::importers::OptionTags::FunctionOfTimeFile>;
```
these option tags need `static constexpr bool pass_metavariables = false;`

nilsdeppe commented on tests/Unit/IO/Importers/Test_ReadSpecThirdOrderPiecewisePolynomial.cpp:
```diff
@@ -0,0 +1,254 @@
+           {"RotationAngle", "RotationAngle"}},
+       std::move(initial_functions_of_time)});
+
+  runner.set_phase(Metavariables::Phase::Testing);
```
we now have non-member functions `ActionTesting::set_phase` that are preferred, same with `next_action`

nilsdeppe reviewed (changes requested):
I'm happy for you to squash immediately
